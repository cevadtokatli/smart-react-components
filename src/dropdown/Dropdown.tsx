import React from "react"
import FixedBox from "../fixed-box/FixedBox"

interface Props {
    children: [JSX.Element, JSX.Element]
    minWidth?: number
    boxShadow?: boolean
    transitionClassName?: string
    transitionType?: string
    transitionDuration?: number
    showAnimation?: boolean
    hideAnimation?: boolean
}

const Dropdown: React.FC<Props> = ({children,minWidth=300,boxShadow,transitionClassName,transitionType,transitionDuration,showAnimation,hideAnimation}) => {
    const [status, setStatus] = React.useState(false)
    return (
        <FixedBox status={status} setStatus={setStatus} minWidth={minWidth} boxShadow={boxShadow} transitionClassName={transitionClassName} transitionType={transitionType} transitionDuration={transitionDuration} showAnimation={showAnimation} hideAnimation={hideAnimation}>
            {children[0]}
            {React.cloneElement(children[1], {setStatus})}
        </FixedBox>
    )
}

export default Dropdown