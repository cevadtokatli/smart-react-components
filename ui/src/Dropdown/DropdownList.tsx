import { PaletteProp, Partial, ResponsiveProp, SetState, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import DropdownListElement from '../components/Dropdown/DropdownListElement'

export interface Props extends Partial<ResponsiveProp<'size', SizeProp>> {
  children: JSX.Element | JSX.Element[]
  hasHover?: boolean
  hasWaveEffect?: boolean
  isOutline?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  waveEffectPalette?: PaletteProp
}

interface PrivateProps {
  setStatus: SetState<boolean>
}

const DropdownList: React.FC<Props> = ({ children, hasHover = true, hasWaveEffect = true, isOutline = true, isSoft, palette = 'primary', setStatus, size = 'medium', sizeSm, sizeMd, sizeLg, sizeXl, waveEffectPalette = 'light' }: Props & PrivateProps) => (
  <DropdownListElement
    dropdownListSize={size}
    dropdownListSizeSm={sizeSm}
    dropdownListSizeMd={sizeMd}
    dropdownListSizeLg={sizeLg}
    dropdownListSizeXl={sizeXl}
    isOutline={isOutline}
    isSoft={isSoft}
    palette={palette}
  >
    { (Array.isArray(children) ? children : [children]).map((item, idx) => item && React.cloneElement(item, { key: item.key ?? idx, hasHover, hasWaveEffect, isOutline, isSoft, palette, setStatus, waveEffectPalette })) }
  </DropdownListElement>
)

export default DropdownList
