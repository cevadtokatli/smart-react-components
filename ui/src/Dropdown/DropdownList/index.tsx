import { PaletteProp, Partial, ResponsiveProp, SetState, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import DropdownListElement from './DropdownListElement'

export interface Props extends Partial<ResponsiveProp<'size', SizeProp>> {
  children: JSX.Element[]
  hasHover?: boolean
  hasWaveEffect?: boolean
  isFilled?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  waveEffectPalette?: PaletteProp
}

interface PrivateProps {
  setStatus: SetState<boolean>
}

const Dropdown: React.FC<Props> = ({ children, hasHover = true, hasWaveEffect = true, isFilled, isSoft, palette = 'primary', setStatus, size = 'medium', sizeSm, sizeMd, sizeLg, sizeXl, waveEffectPalette = 'light' }: Props & PrivateProps) => (
  <DropdownListElement
    dropdownListSize={size}
    dropdownListSizeSm={sizeSm}
    dropdownListSizeMd={sizeMd}
    dropdownListSizeLg={sizeLg}
    dropdownListSizeXl={sizeXl}
    isFilled={isFilled}
    isSoft={isSoft}
    palette={palette}
  >
    { children.map((item, index) => React.cloneElement(item, { key: item.key ?? index, hasHover, hasWaveEffect, isFilled, isSoft, palette, setStatus, waveEffectPalette })) }
  </DropdownListElement>
)

export default Dropdown
