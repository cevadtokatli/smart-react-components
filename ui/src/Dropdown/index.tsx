
import React from 'react'
import FixedBox, { Props, TriggerInteraction } from '../FixedBox'

export * from '../FixedBox'

const Dropdown: React.FC<Props> = ({ afterHide, afterShow, beforeHide, beforeShow, breakpoint, children, elementProps = {}, hasHideAnimation, hasShowAnimation, isDismissible, maxWidth, minWidth, position, setStatus, space, status, transitionClassName, transitionDuration, triggerInteraction }) => {
  const triggerEl = React.useRef<HTMLElement>(null)

  const [localStatus, setLocalStatus] = React.useState(() => false)

  const getTriggerEl = () => ((children[0] as any).ref ?? triggerEl).current as HTMLElement

  React.useEffect(() => {
    getTriggerEl().dispatchEvent(new Event('src.fixedBox.setPosition'))
  }, [children?.[1]?.props?.children?.length])

  return (
    <FixedBox
      afterHide={afterHide}
      afterShow={afterShow}
      beforeHide={beforeHide}
      beforeShow={beforeShow}
      breakpoint={breakpoint}
      elementProps={{
        overflow: ['initial', true],
        ...elementProps,
      }}
      hasHideAnimation={hasHideAnimation}
      hasShowAnimation={hasShowAnimation}
      isDismissible={isDismissible}
      maxWidth={maxWidth}
      minWidth={minWidth}
      position={position}
      setStatus={setStatus ?? setLocalStatus}
      space={space}
      status={status ?? localStatus}
      transitionClassName={transitionClassName}
      transitionDuration={transitionDuration}
      triggerInteraction={triggerInteraction}
    >
      { React.cloneElement(children[0], { ref: (children[0] as any).ref ?? triggerEl, cursor: triggerInteraction & TriggerInteraction.CLICK ? '$cursor.dropdown' : undefined }) }
      { React.cloneElement(children[1], { setStatus: setStatus ?? setLocalStatus }) }
    </FixedBox>
  )
}

export default Dropdown
