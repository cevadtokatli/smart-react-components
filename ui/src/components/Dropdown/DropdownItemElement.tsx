import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp } from '@smart-react-components/core/types'
import styled from 'styled-components'

interface Props extends StyledProps {
  hasHover: boolean
  isActive: boolean
  isDisabled: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
}

export default styled(Div)<Props>(({ theme, hasHover, isActive, isDisabled, isOutline, isSoft, palette }: Props) => `
  border-radius: ${theme.$.radius.dropdown};
  cursor: default;
  transition: 100ms 0s ease-in-out;
  transition-property: background, color, fill;

  ${!isActive
    ? `
      ${hasHover && !theme.$.vars.isMobile
        ? `
          &:hover {
            ${!isSoft
              ? `
                background: ${!isOutline ? theme.$.palette[palette].dynamic : theme.$.palette[palette].main};
                color: ${theme.$.palette[palette].font};
              `
              : `
                background: ${!isOutline ? theme.$.palette[palette].softDark : theme.$.palette[palette].soft};
                color: ${theme.$.palette[palette].softFont};
              `
            }
          }
        `
        : ''
      }
    `
    : `
      ${!isSoft
        ? `
          background: ${!isOutline ? theme.$.palette[palette].dynamicest : theme.$.palette[palette].dynamicer};
          color: ${theme.$.palette[palette].font};
        `
        : `
          background: ${!isOutline ? theme.$.palette[palette].softDarkest : theme.$.palette[palette].softDarker};
          color: ${theme.$.palette[palette].softFont};
        `
      }
    `
  }

  ${isDisabled
    ? `
      opacity: ${theme.$.opacity.dropdownDisabled};
      pointer-events: none;
    `
    : ''
  }
`)
