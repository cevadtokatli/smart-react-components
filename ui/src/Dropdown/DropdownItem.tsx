import { ContentElement, PaletteProp, SetState } from '@smart-react-components/core/types'
import React from 'react'
import WaveEffect from '../WaveEffect'
import DropdownItemElement from '../components/Dropdown/DropdownItemElement'

export interface Props {
  children: ContentElement
  isDisabled?: boolean
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
}

export interface PrivateProps {
  hasHover: boolean
  hasWaveEffect: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  setStatus: SetState<boolean>
  waveEffectPalette: PaletteProp
}

const DropdownItem: React.FC<Props> = ({ children, hasHover, hasWaveEffect, isDisabled, isOutline, isSoft, onClick, palette, setStatus, waveEffectPalette }: (Props & PrivateProps)) => {
  const isClicked = React.useRef(false)

  const handleOnClick = (e: React.MouseEvent<HTMLElement>) => {
    if (isClicked.current || isDisabled) {
      return
    }

    isClicked.current = true

    onClick?.(e)

    if (!e.defaultPrevented) {
      setStatus(false)
    } else {
      isClicked.current = false
    }
  }

  const content = (
    <DropdownItemElement
      cursor="$cursor.dropdown"
      hasHover={hasHover}
      isActive={false}
      isDisabled={isDisabled}
      isOutline={isOutline}
      isSoft={isSoft}
      palette={palette}
      onClick={handleOnClick}
    >
      {children}
    </DropdownItemElement>
  )

  return hasWaveEffect ? <WaveEffect palette={waveEffectPalette}>{content}</WaveEffect> : content
}

export default DropdownItem
