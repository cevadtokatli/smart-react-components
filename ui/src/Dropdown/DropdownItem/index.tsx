import { SetState } from '@smart-react-components/core/types'
import { ContentElement, PaletteProp } from '@smart-react-components/core/types'
import React from 'react'
import WaveEffect from '../../WaveEffect'
import DropdownItemElement from './DropdownItemElement'

export interface Props {
  children: ContentElement
  isDisabled?: boolean
  onClick?: (e: React.MouseEvent<HTMLElement>) => void | Promise<void>
}

interface PrivateProps {
  hasHover: boolean
  hasWaveEffect: boolean
  isFilled: boolean
  isSoft: boolean
  palette: PaletteProp
  setStatus: SetState<boolean>
  waveEffectPalette: PaletteProp
}

const DropdownItem: React.FC<Props> = ({ children, hasHover, hasWaveEffect, isDisabled, isFilled, isSoft, onClick, palette, setStatus, waveEffectPalette }: (Props & PrivateProps)) => {
  const handleOnClick = async (e: React.MouseEvent<HTMLElement>) => {
    if (isDisabled) {
      return
    }

    await onClick?.(e)

    if (!e.defaultPrevented) {
      setStatus(false)
    }
  }

  const content = (
    <DropdownItemElement
      hasHover={hasHover}
      isActive={false}
      isFilled={isFilled}
      isDisabled={isDisabled}
      isSoft={isSoft}
      palette={palette}
      onClick={handleOnClick}
    >
      {children}
    </DropdownItemElement>
  )

  return hasWaveEffect ? <WaveEffect palette={waveEffectPalette}>{content}</WaveEffect> : content
}

export default DropdownItem
