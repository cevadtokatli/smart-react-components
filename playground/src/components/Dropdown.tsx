import React from 'react'
import Dropdown from '@smart-react-components/ui/Dropdown'
import DropdownList from '@smart-react-components/ui/Dropdown/DropdownList'
import DropdownItem from '@smart-react-components/ui/Dropdown/DropdownItem'
import DropdownCategory from '@smart-react-components/ui/Dropdown/DropdownCategory'
import { DropdownArrowIcon } from '@smart-react-components/ui'

export const children = (
  <DropdownCategory label="Category">
    <DropdownItem key="item-1">Item 1</DropdownItem>
    <DropdownItem key="item-2">Item 2</DropdownItem>
    <DropdownItem key="item-3">Item 3</DropdownItem>
  </DropdownCategory>
)

export const Palette = () => (
  <>
    <Dropdown>
      <button>Primary</button>
      <DropdownList>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Secondary</button>
      <DropdownList palette="secondary">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Success</button>
      <DropdownList palette="success">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Danger</button>
      <DropdownList palette="danger">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Warning</button>
      <DropdownList palette="warning">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Info</button>
      <DropdownList palette="info">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Dynamic</button>
      <DropdownList palette="dynamic">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>!Dynamic</button>
      <DropdownList palette="!dynamic">{children}</DropdownList>
    </Dropdown>
  </>
)

export const Outline = () => (
  <>
    <Dropdown>
      <button>Primary</button>
      <DropdownList isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Secondary</button>
      <DropdownList palette="secondary" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Success</button>
      <DropdownList palette="success" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Danger</button>
      <DropdownList palette="danger" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Warning</button>
      <DropdownList palette="warning" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Info</button>
      <DropdownList palette="info" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Dynamic</button>
      <DropdownList palette="dynamic" isOutline={false}>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>!Dynamic</button>
      <DropdownList palette="!dynamic" isOutline={false}>{children}</DropdownList>
    </Dropdown>
  </>
)

export const Soft = () => (
  <>
    <Dropdown>
      <button>Primary</button>
      <DropdownList isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Secondary</button>
      <DropdownList palette="secondary" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Success</button>
      <DropdownList palette="success" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Danger</button>
      <DropdownList palette="danger" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Warning</button>
      <DropdownList palette="warning" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Info</button>
      <DropdownList palette="info" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Dynamic</button>
      <DropdownList palette="dynamic" isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>!Dynamic</button>
      <DropdownList palette="!dynamic" isSoft>{children}</DropdownList>
    </Dropdown>
  </>
)

export const SoftOutline = () => (
  <>
    <Dropdown>
      <button>Primary</button>
      <DropdownList isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Secondary</button>
      <DropdownList palette="secondary" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Success</button>
      <DropdownList palette="success" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Danger</button>
      <DropdownList palette="danger" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Warning</button>
      <DropdownList palette="warning" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Info</button>
      <DropdownList palette="info" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Dynamic</button>
      <DropdownList palette="dynamic" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>!Dynamic</button>
      <DropdownList palette="!dynamic" isOutline={false} isSoft>{children}</DropdownList>
    </Dropdown>
  </>
)

export const Size = () => (
  <>
    <Dropdown>
      <button>Small</button>
      <DropdownList size="small">{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <div>Medium <DropdownArrowIcon height={14} width={14} /></div>
      <DropdownList>{children}</DropdownList>
    </Dropdown>
    <Dropdown>
      <button>Large</button>
      <DropdownList size="large">{children}</DropdownList>
    </Dropdown>
  </>
)
