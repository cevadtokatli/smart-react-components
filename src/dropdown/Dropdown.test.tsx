import React from "react"
import renderTheme from "../test/renderTheme"
import Dropdown from "./Dropdown"
import DropdownList from "./List"
import DropdownItem from "./Item"

describe("<Dropdown />", () => {
    it("should match snapshot", () => {
        const wrapper = renderTheme(
        <Dropdown>
            <button>Toggle</button>
            <DropdownList>
                <DropdownItem>Item 1</DropdownItem>
                <DropdownItem>Item 2</DropdownItem>
                <DropdownItem>Item 3</DropdownItem>
            </DropdownList>
        </Dropdown>)
        expect(wrapper).toMatchSnapshot()
    })
})