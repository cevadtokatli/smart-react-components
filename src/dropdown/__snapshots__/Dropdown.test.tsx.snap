// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should match snapshot 1`] = `
<button>
  Toggle
</button>
`;
